File: src/vectorpin/hash.py
# Copyright 2025 Jascha Wanger / Tarnover, LLC
# SPDX-License-Identifier: Apache-2.0
"""Canonical hashing for embeddings, source content, and metadata.

The protocol is built on three reproducible operations:

  1. canonical_vector_bytes(vector, dtype) -> bytes
     Produces a byte sequence that two implementations will agree on for
     the same float32/float64 array.

  2. hash_vector(vector, dtype) -> "sha256:<hex>"
     SHA-256 over the canonical bytes.

  3. hash_text(text) -> "sha256:<hex>"
     SHA-256 over UTF-8 NFC-normalized text bytes.

These are the only places where bytes get mixed with semantics. Every
attestation field is one of: an explicit string, a hash, a timestamp, or
a signature. No implicit serialization.
"""

from __future__ import annotations

import hashlib
import unicodedata

import numpy as np

# ---- vector hashing ----

CanonicalDtype = str  # "f32" | "f64"


def _np_dtype(dtype: CanonicalDtype) -> np.dtype:
    if dtype == "f32":
        return np.dtype("<f4")  # little-endian float32
    if dtype == "f64":
        return np.dtype("<f8")  # little-endian float64
    raise ValueError(f"unsupported canonical dtype: {dtype!r}")


def canonical_vector_bytes(vector: np.ndarray, dtype: CanonicalDtype = "f32") -> bytes:
    """Return reproducible bytes for an embedding vector.

    Always little-endian, always 1-D. The dtype is fixed by the caller so
    two implementations reach the same byte sequence regardless of how the
    array was constructed.
    """
    if vector.ndim != 1:
        raise ValueError(f"expected 1-D vector, got shape {vector.shape}")
    target = _np_dtype(dtype)
    arr = np.ascontiguousarray(vector.astype(target, copy=False))
    return arr.tobytes()


def hash_vector(vector: np.ndarray, dtype: CanonicalDtype = "f32") -> str:
    """Return the canonical hash of an embedding vector as 'sha256:<hex>'."""
    digest = hashlib.sha256(canonical_vector_bytes(vector, dtype)).hexdigest()
    return f"sha256:{digest}"


# ---- text hashing ----


def hash_text(text: str) -> str:
    """Return the canonical hash of a source text chunk as 'sha256:<hex>'.

    Text is normalized to Unicode NFC then encoded as UTF-8 before hashing.
    NFC chosen because it is the form the W3C and most embedding model
    tokenizers expect; this prevents trivial false-mismatches from upstream
    normalization differences.
    """
    nfc = unicodedata.normalize("NFC", text)
    digest = hashlib.sha256(nfc.encode("utf-8")).hexdigest()
    return f"sha256:{digest}"


def hash_bytes(data: bytes) -> str:
    """SHA-256 over raw bytes, returned as 'sha256:<hex>'."""
    return f"sha256:{hashlib.sha256(data).hexdigest()}"

File: typescript/src/hash.ts
// Copyright 2025 Jascha Wanger / Tarnover, LLC
// SPDX-License-Identifier: Apache-2.0
//
// Canonical hashing for source text and embedding vectors.
//
// These three operations are the only places in the protocol where
// semantic content gets turned into bytes. The Python and Rust ports
// agree on the byte semantics here, so this implementation pins down
// the same contract:
//
//   * Vectors: little-endian, 1-D, packed `f32` or `f64` bytes.
//   * Text: UTF-8 of the NFC-normalized string.
//   * Output digests: prefixed with `"sha256:"` and lowercase hex.

import { sha256 } from '@noble/hashes/sha2';
import { bytesToHex } from '@noble/hashes/utils';

/** Canonical scalar dtype identifier carried in the wire format. */
export type VecDtype = 'f32' | 'f64';

/**
 * A 1-D vector view accepted by the canonicalization helpers. Plain
 * `number[]`, `Float32Array`, and `Float64Array` are all supported
 * — the dtype the caller passes determines how the bytes are packed.
 */
export type VectorInput = readonly number[] | Float32Array | Float64Array;

/**
 * Reproducible byte form of an embedding vector.
 *
 * Always little-endian, always packed, always under the dtype the
 * caller specifies. Two implementations must agree on these bytes
 * byte-for-byte for cross-language verification to work.
 */
export function canonicalVectorBytes(vector: VectorInput, dtype: VecDtype): Uint8Array {
  if (vector.length === 0) {
    throw new Error('cannot canonicalize empty vector');
  }
  const bytesPerElem = dtype === 'f32' ? 4 : 8;
  const out = new Uint8Array(vector.length * bytesPerElem);
  const view = new DataView(out.buffer, out.byteOffset, out.byteLength);
  if (dtype === 'f32') {
    for (let i = 0; i < vector.length; i++) {
      // `vector[i]!` is safe — we just bounded i to vector.length.
      view.setFloat32(i * 4, vector[i]!, /* littleEndian */ true);
    }
  } else {
    for (let i = 0; i < vector.length; i++) {
      view.setFloat64(i * 8, vector[i]!, /* littleEndian */ true);
    }
  }
  return out;
}

/** SHA-256 of a vector's canonical bytes, formatted as `"sha256:<hex>"`. */
export function hashVector(vector: VectorInput, dtype: VecDtype = 'f32'): string {
  return sha256Prefixed(canonicalVectorBytes(vector, dtype));
}

/** SHA-256 of a string after Unicode NFC normalization and UTF-8 encoding. */
export function hashText(text: string): string {
  const nfc = text.normalize('NFC');
  return sha256Prefixed(new TextEncoder().encode(nfc));
}

/** SHA-256 over arbitrary bytes, formatted as `"sha256:<hex>"`. */
export function hashBytes(data: Uint8Array): string {
  return sha256Prefixed(data);
}

function sha256Prefixed(data: Uint8Array): string {
  return `sha256:${bytesToHex(sha256(data))}`;
}
